src/components/ShareURL.tsx

```diff
@@ -0,0 +1,93 @@
+import { t } from 'ttag'
+import React, { useState } from 'react'
+import { CopyToClipboard } from 'react-copy-to-clipboard'
+import { useDispatch } from 'react-redux'
+import ModalCard from './ModalCard'
+import config from '../config'
+import { dumpConfiguration } from '../actions/saves'
+import { post } from '../io'
+import { setError } from '../actions/error'
+
+type ShareURLProps = {
+  configuration: any
+  version: number
+}
+
+const ShareURL = ({ configuration, version }: ShareURLProps) => {
+  const dispatch = useDispatch()
+  const [url, setUrl] = useState('')
+  const [fetchingUrl, setFetchingUrl] = useState(false)
+  const [urlCopied, setUrlCopied] = useState(false)
+
+  const resetState = () => {
+    setUrl('')
+    setFetchingUrl(false)
+    setUrlCopied(false)
+  }
+
+  const onFetchURL = () => {
+    resetState()
+    setFetchingUrl(true)
+    const shareUrl = `${config.apiRoot}share-urls/`
+    const data = {
+      configuration: JSON.stringify(dumpConfiguration(configuration)),
+      version,
+    }
+    post(shareUrl, data)
+      .then(response => {
+        const { status } = response
+        if (status < 200 || status >= 300) {
+          throw new Error(`There was an unexpected error creating the share URL. Status: ${status}`)
+        } else {
+          response.json().then(shareURL => {
+            const { hash } = shareURL
+            const { protocol, host, pathname } = window.location
+            setUrl(`${protocol}//${host + pathname}?s=${hash}`)
+          })
+        }
+      })
+      .catch(e => {
+        dispatch(setError('Error', 'There was a problem creating your URL.', e.message))
+      })
+      .finally(() => {
+        setFetchingUrl(false)
+      })
+  }
+
+  return (
+    <>
+      <ModalCard
+        onHide={() => !fetchingUrl && resetState()}
```

Comment:
Only reset the state if not fetching the URL?

Reply:
`onHide` is run when the close modal "x" is clicked. Without `!fetchingUrl &&`, if the user clicks the "x" before the url has loaded, then url state will trickle in after state has reset, causing the modal to reopen. If they close it and try to load a different url there will be two modals.

Reply:
Gotcha.